packages/go_router/lib/src/router.dart

```diff
@@ -302,12 +305,64 @@ class GoRouter extends ChangeNotifier implements RouterConfig<RouteMatchList> {
   /// Navigate to a URI location w/ optional query parameters, e.g.
   /// `/family/f2/person/p1?color=blue`
   void go(String location, {Object? extra}) {
-    log.info('going to $location');
-    routeInformationProvider.go(location, extra: extra);
+    if (routerDelegate.isOnExitBeingCalled) {
+      log.info('ignore go as onExit is being called');
+      return;
+    }
+    routerDelegate.isOnExitBeingCalled = true;
+    final RouteInformation routeInformation = RouteInformation(
+      // TODO(chunhtai): remove this ignore and migrate the code
+      // https://github.com/flutter/flutter/issues/124045.
+      // ignore: deprecated_member_use
+      location: location,
+      state: RouteInformationState<void>(
+        extra: extra,
+        type: NavigatingType.go,
+      ),
+    );
+    final BuildContext context = configuration.navigatorKey.currentContext!;
+    routeInformationParser
+        .parseRouteInformationWithDependencies(routeInformation, context)
+        .then((RouteMatchList newMatchList) async {
```

Comment:
These onExit logic should be in the GoRouteInformationParser.

GoRouter will pass the go/push to routeinformationprovider as usual, It would be  GoRouterInformationParser to decide how to produce a final RouteMatchList.